### PR TITLE
infra for algebraic simplification and simplifications for Add and Mu…

### PR DESCRIPTION
**Pull request by Krovatkin**
…ltiply (including broadcast consts)

### CODE REVIEW

File: test/algebraic_simplification.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
+/*******************************************************************************
+* Copyright 2017-2018 Intel Corporation
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply test/algebraic_simplification.cpp]: will do

File: src/ngraph/pass/algebraic_simplification.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+template <typename T>
+static std::shared_ptr<pattern::Matcher>
+    create_binary_matcher(std::shared_ptr<pattern::op::Label> label,
+                          std::shared_ptr<pattern::op::Label> const_label)
+{
+    auto bcst_pred = [](std::shared_ptr<Node> n) {
+        return std::dynamic_pointer_cast<op::Broadcast>(n) != nullptr;
+    };
+    auto bcst = std::make_shared<pattern::op::Any>(const_label, bcst_pred);
```
[the remark this answers is not in the document]
Comment (mbrookhart) [reply src/ngraph/pass/algebraic_simplification.cpp]: I see #891 and walk back my statement, this isn't actually an Any class.

[the remark this answers is not in the document]
Comment (Krovatkin) [reply src/ngraph/pass/algebraic_simplification.cpp]: 😄

File: src/ngraph/pass/algebraic_simplification.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    return matcher;
+}
+
+static bool simplify_multiply(std::shared_ptr<Node> n)
+{
+    NGRAPH_DEBUG << "In simplify_multiply for " << n->get_name();
+    auto iconst = ngraph::make_zero(element::i32, Shape{});
+    auto label = std::make_shared<pattern::op::Label>(iconst);
+    auto const_label = std::make_shared<pattern::op::Label>(iconst, nullptr, NodeVector{iconst});
+    auto matcher = create_binary_matcher<op::Multiply>(label, const_label);
```
[the remark this answers is not in the document]
Comment (Krovatkin) [reply src/ngraph/pass/algebraic_simplification.cpp]: yes, `test/pattern.cpp` contains some tests for commutativeness of `+` and `*`.
Here's an example: https://github.com/NervanaSystems/ngraph/blob/master/test/pattern.cpp#L430

But if you feel we need to add more commutativeness tests for `AlgebraicSimplifier` I can add some.

**Comment by Krovatkin** [comment]
@diyessi @mbrookhart ping...

**Pull request merged**